### PR TITLE
Import rails-bin linter to rails internal tools lib

### PR DESCRIPTION
**Pull request by zzak**
Alternative to #47813 to import the linter files directly to Rails source tree

```bash
$ zzak@mbp16 rails % ruby -v
ruby 3.2.2 (2023-03-30 revision e51014f9c0) [arm64-darwin22]



$ zzak@mbp16 rails % bundle exec tools/check-changelogs .
.............

13 changelogs inspected, 0 offenses detected



$ zzak@mbp16 rails % bundle exec tools/check-config-docs .



$ zzak@mbp16 rails % bundle exec tools/check-links .
https://edgeapi.rubyonrails.org/classes/ActiveSupport/MessageEncryptor.html#method-i-encrypt_and_sign
```

### CODE REVIEW

**Review by skipkayhil — COMMENTED** [review]
I _think_ I'd prefer importing `rail_inspector` just because the scripts were all combined into the single Thor entrypoint. If you think `check-links` is valuable I can add that to `rail_inspector` as that is currently missing. I think `regen-dummy` _could_ also be worth importing, but it seems less important due to Rafael's comment: https://github.com/rails/rails/pull/47259#issuecomment-1451005844

Something else that probably needs to be done is better namespacing of the modules. I'm realizing now that these are all in the global namespace so it may be better to nest them when running in the context of a larger Gemfile. (I've opened an issue for that here: skipkayhil/rail_inspector#2)

Edit: I'm also noticing that this fails Rails' rubocop. Opened an issue for that too: skipkayhil/rail_inspector#3

Ah something else: I would be disappointed if we lose the changelog tests. They've been really useful in ensuring changes don't regress. With that being said, I don't know what running those tests in CI here would look like. I'm assuming it would be a Github Action and they should also probably only run if these files are being changed

**Comment by skipkayhil** [comment]
I opened some example PRs on my fork to show that the `rail_inspector.yml` workflow will only run if the folder is touched:
skipkayhil/rails#3 <- no change, no `rail_inspector` workflow
skipkayhil/rails#4 <- folder has change, `rail_inspector` workflow is run

File: Gemfile [hunk 1]
```diff
@@ -40,6 +40,10 @@ gem "json", ">= 2.0.0"
 # Workaround until Ruby ships with cgi version 0.3.6 or higher.
 gem "cgi", ">= 0.3.6", require: false
 
+group :lint do
```
Comment (zzak) [inline Gemfile:43]: If we add this group to the gemfile, does it get required automatically? Or do we need to add `required: false` here?

File: tools/rail_inspector/README.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+## Installation
+
+Install the gem and add to the application's Gemfile by executing:
+
+    $ bundle add rail_inspector
+
+If bundler is not being used to manage dependencies, install the gem by
+executing:
+
+    $ gem install rail_inspector
```
Comment (zzak) [inline tools/rail_inspector/README.md:14]: Do we want to still release it as a gem? 🤔

File: tools/rail_inspector/README.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+To install this gem onto your local machine, run `bundle exec rake install`. To
+release a new version, update the version number in `version.rb`, and then run
+`bundle exec rake release`, which will create a git tag for the version, push
+git commits and the created tag, and push the `.gem` file to
+[rubygems.org](https://rubygems.org).
+
+## Contributing
+
+Bug reports and pull requests are welcome on GitHub at
+https://github.com/skipkayhil/rail_inspector.
```
Comment (zzak) [inline tools/rail_inspector/README.md:41]: Hopefully there are no bug reports, but they should probably go to rails/rails?

File: tools/rail_inspector/lib/rail_inspector.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+module RailInspector
+end
```
Comment (zzak) [inline tools/rail_inspector/lib/rail_inspector.rb:28]: Do these files need a newline at the end? We ran rails rubocop on them right? 🤔

File: Gemfile (hunk 1 above)
Comment (skipkayhil) [reply Gemfile:43]: Updated with `require: false` :+1:

File: tools/rail_inspector/lib/rail_inspector.rb (hunk 4 above)
Comment (skipkayhil) [reply tools/rail_inspector/lib/rail_inspector.rb:28]: Yes, `rubocop` should be running on them. As far as I know there is a newline at the end of this file or Github would put the red icon indicating that there is not

File: tools/rail_inspector/README.md (hunk 2 above)
Comment (skipkayhil) [reply tools/rail_inspector/README.md]: Removed README because I don't think it was adding much

File: .github/workflows/rail_inspector.yml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+name: Rail Inspector
```
Comment (zzak) [inline .github/workflows/rail_inspector.yml:1]: I'm wondering if we need to add this workflow if we're going to move this to buildkite soon anyways 🤔

**Pull request merged**